README.md
PikadayAngular (with Angular4)
====================================

### A Dart Datepicker usable as a AngularDart component

* uses [Pikaday] dart-wrapper of a lightweight js-lib with datepicker functionality ([Pikaday.js], less than 5kb minified and gzipped)
* one optinal dependency, if date format needs to be changed ([Moment.js], less than 25kb minified and gzipped)
* modular CSS classes for easy styling

![screenshot]

If you do find bugs or have feature requests please submit them to the [issues]
Also see the [changelog](CHANGELOG.md)


## Usage

### html

Link to the [Pikaday.js]-js library (with it's css file) and
the optional but highly advisable [Moment.js] library (for advanced date formating) in your index.html-file.
Currently, you also need to import the pikaday_dart_helpers.js file from the [Pikaday] package:

Assuming you put these files into /web/jsLibs/ the imports look like this:
```html
<link rel="stylesheet" href="jsLibs/pikaday.css">
<script src="jsLibs/moment.js"></script>
<script src="jsLibs/pikaday.js"></script>
<script src="packages/pikaday/pikaday_dart_helpers.js"></script>
```

Download a local copy of those libs, put them in or under the directory your index.html file is in,
an link accordingly.

### Angular

Check out the example in web/main.dart!

Import and use **PikadayComponent** into your AngularDart component:

```dart
import 'package:pikaday_datepicker_angular/pikaday_datepicker_angular.dart';

// example of how to use the PikadayComponent to display a day picker.
@Component(
    selector: 'pick-day-component',
    template: '''<pikaday [(day)]="selectedDay" format="YYYY-MM-DD"
                          placeholder="select a day">
                 </pikaday>
                 <div>day: {{selectedDay | date}}</div>''',
    directives: const [PikadayComponent],
    pipes: const [COMMON_PIPES])
class DayPickerComponent {
  DateTime selectedDay = new DateTime(2015, 2, 1);
}

// example of how to use the PikadayComponent to display a day and time picker.
@Component(
    selector: 'pick-day-time-component',
    template: '''<pikaday [(day)]="selectedDayTime" format="YYYY-MM-DD h:mm A"
                          firstDay="1" minDate="2010-1-1" maxDate="2025-12-31"
                          [showTime]="true" [autoClose]="false" [incrementMinuteBy]="5"
                          placeholder="select a day and time">
                 </pikaday>
                 <div>day: {{selectedDayTime | date}}</div>
                 <div>time: {{daytime24}} o'clock</div>''',
    directives: const [PikadayComponent],
    pipes: const [COMMON_PIPES])
class DayTimePickerComponent {
  String get daytime24 =>
      "${selectedDayTime.hour}:${_padTimeAs2Chars(selectedDayTime.minute)}";
  DateTime selectedDayTime = new DateTime(2015, 2, 1, 13, 30);

  String _padTimeAs2Chars(int hoursOrMinutes) =>
      hoursOrMinutes < 10 ? "0$hoursOrMinutes" : hoursOrMinutes.toString();
}
```

If you want to compile the example under web to js, you have to change the transformer configuration to
```
transformers:
### (default) configuration if this package should be used as a library (import) or if the app should be execute as Dart in Dartium
#- angular
### configuration if example web folder should be compiled to js
- angular:
    entry_points: web/main.dart
- dart_to_js_script_rewriter
```

### Basic/Vanilla Dart Usage

If you're looking for a datepicker within a plain Dart project (not with angular),
check out the [Pikaday] package.

### Angular2 component - Configuration

PikadayComponent has many useful options:
- two-way data binding:
  - `day` the DateTime instance to display/update (replacing/combining `defaultDay` and `setDefaultDate` from PikadayParams)
- one-way attributes
  - `cssClasses` setting css classes on input (<input class="{{cssClasses}}>)
  - `placeholder` sets the placeholder of the pikaday-inputfield
- date attributes directly used to initialize PikadayParams
  - `bound` automatically show/hide the datepicker on input field focus (default `true` if `field` is set)
  - `position` preferred position of the datepicker relative to the form field, e.g.: `top right`, `bottom right` **Note:** automatic adjustment may occur to avoid datepicker from being displayed outside the viewport, see (default to 'bottom left')
  - `reposition` can be set to false to not reposition datepicker within the viewport, forcing it to take the configured `position` (default: true)
  - `format` the output format used within the input element (requires [Moment.js] for custom formatting)
  - `formatStrict` the default flag for moment's strict date parsing (requires [Moment.js] for custom formatting)
  - `firstDay` first day of the week (0: Sunday, 1: Monday, etc)
  - `minDate` the minimum/earliest date that can be selected (this should be a native Date object - e.g. `new Date()` or `moment().toDate()`)
  - `maxDate` the maximum/latest date that can be selected (this should be a native Date object - e.g. `new Date()` or `moment().toDate()`)
  - `disableWeekends` disallow selection of Saturdays or Sundays
  - `yearRange` number of years either side (e.g. `10`) or array of upper/lower range (e.g. `[1900,2015]`)
  - `showWeekNumber` show the ISO week number at the head of the row (default `false`)
  - `isRTL` reverse the calendar for right-to-left languages
  - `i18n` language defaults for month and weekday names (see internationalization example on [Pikaday.js])
  - `yearSuffix` additional text to append to the year in the title
  - `showMonthAfterYear` render the month after year in the title (default `false`)
  - `showDaysInNextAndPreviousMonths` render days of the calendar grid that fall in the next or previous months to the current month instead of rendering an empty table cell (default: true)
  - `numberOfMonths` number of visible calendars
  - `mainCalendarIsLeft` when `numberOfMonths` is used, this will help you to choose where the main calendar will be (default `true`/`left`, can be set to `false`/`right`). Only used for the first display or when a selected date is not already visible
  - `theme` define a classname that can be used as a hook for styling different themes (default `null`)
- time attributes directly used to initialize PikadayParams
  - `autoClose` bool or boolish string
  - `use24hour` bool or boolish string
  - `showTime` bool or boolish string
  - `timeLabel` string
  - `showMinutes` bool or boolish string
  - `showSeconds` bool or boolish string
  - `incrementHourBy` num
  - `incrementMinuteBy` num
  - `incrementSecondBy` num

### Common problems

- console error msg: **EXCEPTION: Error: self.Pikaday is not a constructor**
You forgot to import pikaday.js in your html-file.
- console error msg: **TypeError: self.getPikadayMillisecondsSinceEpoch is not a function**
You forgot to import pikaday_dart_helpers.js in your html-file.

```html
<link rel="stylesheet" href="jsLibs/pikaday.css">
<script src="jsLibs/moment.js"></script>
<script src="jsLibs/pikaday.js"></script>
<script src="packages/pikaday/pikaday_dart_helpers.js"></script>
```
* * *

## Authors

* Stephan Schröder [GitHub](https://github.com/simon-void)

## Contributors

* Tobechukwu Osakwe [GitHub](https://github.com/thosakwe)

Thanks to [David Bushell](https://github.com/dbushell) for writing [Pikaday.js].
Thanks to [John Ryan](https://github.com/johnpryan) for writing [Pikaday].

Copyright © 2017 Stephan Schröder | BSD & MIT license

  [Pikaday]:      https://pub.dartlang.org/packages/pikaday
  [Pikaday.js]:   https://github.com/dbushell/Pikaday                              "Pikaday"
  [Moment.js]:    http://momentjs.com/                                             "moment.js"
  [screenshot]:   https://raw.github.com/dbushell/Pikaday/gh-pages/screenshot.png  "Screenshot"
  [issues]:       https://github.com/simon-void/pikaday_component_angular2/issues           "Issues"
  

CHANGELOG.md
# Changelog

## 3.1.0 - 2018-02-27

- moved back to the official, new (1.2) version of pickaday.dart instead of using my fork of it
- merged in time support by thosakwe that used new pickaday.dart options, e.g. [showTime]="true" [incrementMinuteBy]="4"

## 3.0.1 - 2017-10-24

- commented out dart_to_js_script_rewriter from standard transformer section
 (it's only meant to be used if the provided example is to be looked at)

## 3.0.0 - 2017-09-15

- upgraded the pikaday_datepicker_angular2 package version 2.1.4.
from angular2 to angular4


lib/src/pikaday_component.dart
import 'dart:async';
import 'dart:html';

import 'package:angular/angular.dart';
import 'package:js/js.dart';
import 'package:pikaday/pikaday.dart';
import 'package:pikaday/pikaday_dart_helpers.dart';

import 'conversion.dart';

/// AngularDart component wrapper around the Pikaday-js lib. You will have to
/// link to pikaday.js (Get the latest version from it's
/// [GitHub page](https://github.com/dbushell/Pikaday) and if you want
/// a custom date format (which is highly likable) also to [moment.js](http://momentjs.com/)).
///
/// Attribute documentation adapted from the
/// [pikaday.js documentation](https://github.com/dbushell/Pikaday).
///
/// You can't set a container DOM node nore a callback, but you can listen to
/// dayChange to be informed about selected days (DateTime instances).

@Component(
    selector: 'pikaday',
    template:
        '<input type="text" id="{{id}}" class="{{cssClasses}}" placeholder="{{placeholder}}">')
class PikadayComponent implements AfterViewInit {
  static int _componentCounter = 0;
  final String id = "pikadayInput${++_componentCounter}";

  /// css-classes to be set on the pikaday-inputfield via <input class="{{cssClasses}}>
  @Input()
  String cssClasses = "";

  /// Sets the placeholder of the pikaday-inputfield.
  @Input()
  String placeholder;

  Pikaday _pikaday;
  // default options
  final PikadayOptions _options = new PikadayOptions(
      showTime: false,
      autoClose: true,
      use24hour: false,
      incrementMinuteBy: 15,
      incrementSecondBy: 10,
      showSeconds: false);

  bool get _isInitPhase => _pikaday == null;

  /// Emits selected dates (day and time).
  final _dayChange = new StreamController<DateTime>();
  @Output()
  Stream<DateTime> get dayChange => _dayChange.stream;

  /// The initially selected day and time
  /// Combines [PikadayOptions.defaultDate] with [PikadayOptions.setDefaultDate]. 
  @Input()
  void set day(DateTime day) {
    if (_isInitPhase) {
      _options.defaultDate = day;
      _options.setDefaultDate = day != null;
    } else {
      var dayMillies = day?.millisecondsSinceEpoch;
      setPikadayMillisecondsSinceEpoch(_pikaday, dayMillies);
    }
  }

  /// Optional boolean property to specify whether to show time controls with calendar or not.
  /// <bool> or <String>. default: false. Forwards to [PikadayOptions.showTime]. 
  @Input()
  void set showTime(value) {
    _options.showTime = boolValue(value);
  }

  /// Optional numeric property to prevent calendar from auto-closing after date is selected.
  /// <bool> or <String>. default: true. Should probably be false if showTime: true. Forwards to [PikadayOptions.autoClose]. 
  @Input()
  void set autoClose(value) {
    _options.autoClose = boolValue(value);
  }

  /// Optional boolean property to specify whether to use 24 hours format or not.
  /// <bool> or <String>. default: false. Forwards to [PikadayOptions.use24hour]. 
  @Input()
  void set use24hour(value) {
    _options.use24hour = boolValue(value);
  }

  /// Optional string added to left of time select.
  /// Forwards to [PikadayOptions.timeLabel]. 
  @Input()
  void set timeLabel(String value) {
    _options.timeLabel = value;
  }

  /// Optional boolean property to specify whether to show minute controls with calendar or not.  /// <bool> or <String>. default: true. Forwards to [PikadayOptions.showMinutes]. 
  @Input()
  void set showMinutes(value) {
    _options.showMinutes = boolValue(value);
  }

  /// Optional boolean property to specify whether to show second controls with calendar or not.
  /// <bool> or <String>. default: false. Forwards to [PikadayOptions.showSeconds]. 
  @Input()
  void set showSeconds(value) {
    _options.showSeconds = boolValue(value);
  }

  /// default: 1. Forwards to [PikadayOptions.incrementHourBy]. 
  @Input()
  void set incrementHourBy(num value) {
    _options.incrementHourBy = value;
  }

  /// default: 15. Forwards to [PikadayOptions.incrementMinuteBy]. 
  @Input()
  void set incrementMinuteBy(num value) {
    _options.incrementMinuteBy = value;
  }

  /// default: 10. Forwards to [PikadayOptions.incrementSecondBy]. 
  @Input()
  void set incrementSecondBy(num value) {
    _options.incrementSecondBy = value;
  }

  /// Automatically show/hide the datepicker on field focus.
  /// Default: true if field is set.
  /// <bool> or <String>. Forwards to [PikadayOptions.bound]. 
  @Input()
  void set bound(bound) {
    _options.bound = boolValue(bound);
  }

  /// Preferred position of the datepicker relative to the form field
  /// (e.g. 'top right'). Automatic adjustment may occur to avoid
  /// displaying outside the viewport. Default: 'bottom left'.
  /// Forwards to [PikadayOptions.position]. 
  @Input()
  void set position(String position) {
    _options.position = position;
  }

  /// Can be set to false to not reposition the datepicker within the
  /// viewport, forcing it to take the configured position. Default: true.
  /// <bool> or <String>. Forwards to [PikadayOptions.reposition].
  @Input()
  void set reposition(reposition) {
    _options.reposition = boolValue(reposition);
  }

  /// The default output format for toString() and field value.
  /// Requires Moment.js for custom formatting.
  /// Forwards to [PikadayOptions.format].
  @Input()
  void set format(String format) {
    _options.format = format;
  }

  /// First day of the week (0: Sunday, 1: Monday, etc).
  /// <int> or <String>. Forwards to [PikadayOptions.firstDay]. 
  @Input()
  void set firstDay(firstDay) {
    _options.firstDay = intValue(firstDay);
  }

  /// <DateTime> or <String> with format YYYY-MM-DD. Forwards to [PikadayOptions.minDate]. 
  @Input()
  void set minDate(minDate) {
    final minDateAsDateTime = dayValue(minDate);
    if (_isInitPhase) {
      _options.minDate = minDateAsDateTime;
    } else {
      var minDateMillies = minDateAsDateTime?.millisecondsSinceEpoch;
      setPikadayMinDateAsMillisecondsSinceEpoch(_pikaday, minDateMillies);
    }
  }

  /// <DateTime> or <String> with format YYYY-MM-DD. Forwards to [PikadayOptions.maxDate]. 
  @Input()
  void set maxDate(maxDate) {
    final maxDateAsDateTime = dayValue(maxDate);
    if (_isInitPhase) {
      _options.maxDate = maxDateAsDateTime;
    } else {
      var maxDateMillies = maxDateAsDateTime?.millisecondsSinceEpoch;
      setPikadayMaxDateAsMillisecondsSinceEpoch(_pikaday, maxDateMillies);
    }
  }

  /// Disallow selection of Saturdays and Sundays.
  /// Forwards to [PikadayOptions.disableWeekends]. 
  @Input()
  void set disableWeekends(disableWeekends) {
    _options.disableWeekends = boolValue(disableWeekends);
  }

  /// <int>, <List<int>> or <String> (single '1990' or double '1980,2020').
  /// Forwards to [PikadayOptions.yearRange]. 
  @Input()
  void set yearRange(yearRange) {
    _options.yearRange = yearRangeValue(yearRange);
  }

  /// Show the ISO week number at the head of the row. Default: false.
  /// <bool> or <String>. Forwards to [PikadayOptions.showWeekNumber]. 
  @Input()
  void set showWeekNumber(showWeekNumber) {
    _options.showWeekNumber = boolValue(showWeekNumber);
  }

  /// Reverse the calendar for right-to-left languages. Default: false.
  /// <bool> or <String>. Forwards to [PikadayOptions.isRTL]. 
  @Input()
  void set isRTL(isRTL) {
    _options.isRTL = boolValue(isRTL);
  }

  /// Language defaults for month and weekday names.
  /// Forwards to [PikadayOptions.i18n]. 
  @Input()
  void set i18n(PikadayI18nConfig i18n) {
    _options.i18n = i18n;
  }

  /// Additional text to append to the year in the title.
  /// Forwards to [PikadayOptions.yearSuffix]. 
  @Input()
  void set yearSuffix(String yearSuffix) {
    _options.yearSuffix = yearSuffix;
  }

  /// Render the month after the year in the title. Default: false.
  /// <bool> or <String>. Forwards to [PikadayOptions.showMonthAfterYear]. 
  @Input()
  void set showMonthAfterYear(showMonthAfterYear) {
    _options.showMonthAfterYear = boolValue(showMonthAfterYear);
  }

  /// Render days of the calendar grid that fall in the next or previous months to the current month instead of rendering an empty table cell. Default: false.
  /// <bool> or <String>. Forwards to [PikadayOptions.showDaysInNextAndPreviousMonths]. 
  @Input()
  void set showDaysInNextAndPreviousMonths(showDaysInNextAndPreviousMonths) {
    _options.showDaysInNextAndPreviousMonths =
        boolValue(showDaysInNextAndPreviousMonths);
  }

  /// Number of visible calendars.
  /// <int> or <String>. Forwards to [PikadayOptions.numberOfMonths]. 
  @Input()
  void set numberOfMonths(numberOfMonths) {
    _options.numberOfMonths = intValue(numberOfMonths);
  }

  /// When numberOfMonths is used, this will help you to choose where the
  /// main calendar will be (default left, can be set to right). Only used
  /// for the first display or when a selected date is not already visible.
  /// Forwards to [PikadayOptions.mainCalendar]. 
  /// permitted values: "left", "right";
  @Input()
  void set mainCalendar(String mainCalendar) {
    if (mainCalendar == "right" || mainCalendar == "left") {
      _options.mainCalendar = mainCalendar;
    }
    throw new ArgumentError(
        "should only be 'left' or 'right', but was: $mainCalendar");
  }

  /// Define a class name that can be used as a hook for styling different
  /// themes. Default: null.
  /// Forwards to [PikadayOptions.theme]. 
  @Input()
  void set theme(String theme) {
    _options.theme = theme;
  }

  @override
  ngAfterViewInit() {
    _options.field = querySelector('#$id');
    _options.onSelect = allowInterop((dateTimeOrDate) {
      var day = dateTimeOrDate is DateTime
          ? dateTimeOrDate
          : new DateTime.fromMillisecondsSinceEpoch(
              getPikadayMillisecondsSinceEpoch(_pikaday));

      if (day != _options.defaultDate) {
        _options.defaultDate = day;
        _dayChange.add(day);
      }
    });

    _pikaday = new Pikaday(_options);

    // Currently Dart's DateTime is not correctly mapped to JS's Date
    // so they are converted to millies as transferred as int values.
    workaroundDateTimeConversionIssue(
      DateTime day,
      DateTime minDate,
      DateTime maxDate,
    ) {
      if (day != null) {
        var millies = day.millisecondsSinceEpoch;
        setPikadayMillisecondsSinceEpoch(_pikaday, millies);
      }
      if (minDate != null) {
        var millies = minDate.millisecondsSinceEpoch;
        setPikadayMinDateAsMillisecondsSinceEpoch(_pikaday, millies);
      }
      if (maxDate != null) {
        var millies = maxDate.millisecondsSinceEpoch;
        setPikadayMaxDateAsMillisecondsSinceEpoch(_pikaday, millies);
      }
    }

    workaroundDateTimeConversionIssue(
        _options.defaultDate, _options.minDate, _options.maxDate);
  }
}


web/main.dart
import 'package:angular/angular.dart';

import 'package:pikaday_datepicker_angular/pikaday_datepicker_angular.dart';

void main() {
  bootstrap(PickerExampleComponent);
}

// example app to showcase the PikadayComponent.
@Component(
  selector: 'showcase-pikadate-component',
  template: '''<pick-day-component ></pick-day-component><br />
               <pick-day-time-component ></pick-day-time-component>''',
  directives: const [DayTimePickerComponent, DayPickerComponent],
)
class PickerExampleComponent {}

// example of how to use the PikadayComponent to display a day picker.
@Component(
    selector: 'pick-day-component',
    template: '''<pikaday [(day)]="selectedDay" format="YYYY-MM-DD"
                          placeholder="select a day">
                 </pikaday>
                 <div>day: {{selectedDay | date}}</div>''',
    directives: const [PikadayComponent],
    pipes: const [COMMON_PIPES])
class DayPickerComponent {
  DateTime selectedDay = new DateTime(2015, 2, 1);
}

// example of how to use the PikadayComponent to display a day and time picker.
@Component(
    selector: 'pick-day-time-component',
    template: '''<pikaday [(day)]="selectedDayTime" format="YYYY-MM-DD h:mm A"
                          firstDay="1" minDate="2010-1-1" maxDate="2025-12-31"
                          [showTime]="true" [autoClose]="false" [incrementMinuteBy]="5"
                          placeholder="select a day and time">
                 </pikaday>
                 <div>day: {{selectedDayTime | date}}</div>
                 <div>time: {{daytime24}} o'clock</div>''',
    directives: const [PikadayComponent],
    pipes: const [COMMON_PIPES])
class DayTimePickerComponent {
  String get daytime24 =>
      "${selectedDayTime.hour}:${_padTimeAs2Chars(selectedDayTime.minute)}";
  DateTime selectedDayTime = new DateTime(2015, 2, 1, 13, 30);

  String _padTimeAs2Chars(int hoursOrMinutes) =>
      hoursOrMinutes < 10 ? "0$hoursOrMinutes" : hoursOrMinutes.toString();
}
